File: c/whilei.c
#include "interpret.h"
#include "lexer.h"
#include "parser.h"
#include "util.h"
#include "virtual_stack.h"

#include <stdio.h>
#include <string.h>

int main(int argc, char **argv) {
    int trace = 0;
    const char *srcpath = NULL;
    for (int i = 1; i < argc; i++) {
        if (!strcmp(argv[i], "--trace"))
            trace = 1;
        else if (argv[i][0] == '-') {
            fprintf(stderr, "[usage] %s [--trace] file.while\n", argv[0]);
            return 1;
        } else
            srcpath = argv[i];
    }
    if (!srcpath) {
        fprintf(stderr, "[usage] %s [--trace] file.while\n", argv[0]);
        return 1;
    }
    char *src = xread_file(srcpath);
    Stmt *tree = parse(tokenize(src));
    clear_todo();
    Code code = compile_stack(tree);
    const char *todo = take_todo();
    if (todo) {
        fprintf(stderr, "error: %s\n", todo);
        return 1;
    }
    interpret(code, 0, trace);
    return 0;
}

File: python/whilei.py
#!/usr/bin/env python3
"""仮想スタック機械で WhileLang プログラムを実行する。

    python whilei.py ../test/assign.while
    python whilei.py --trace ../test/assign.while
"""

from __future__ import annotations

import sys
from pathlib import Path

from interpret import interpret
from lexer import tokenize
from parser import parse
from virtual_stack import compile_stack


def main(argv: list[str] | None = None) -> int:
    args = sys.argv[1:] if argv is None else argv
    trace = False
    if args and args[0] == "--trace":
        trace = True
        args = args[1:]
    if len(args) != 1:
        print(f"[usage] {Path(sys.argv[0]).name} [--trace] file.while", file=sys.stderr)
        return 1
    path = Path(args[0])
    if not path.exists():
        print(f"ファイルがありません: {path}", file=sys.stderr)
        return 1
    try:
        ast = parse(tokenize(path.read_text(encoding="utf-8")))
        interpret(compile_stack(ast), trace=trace)
    except Exception as e:
        print(f"error: {e}", file=sys.stderr)
        return 1
    return 0


if __name__ == "__main__":
    sys.exit(main())
